File: tools/ci/license_check.py
#!/usr/bin/env python3

import zipfile
from os import path as osp
from xml.etree.ElementTree import parse

XML_HEADER = '{http://maven.apache.org/POM/4.0.0}'


# Extract CDC revision from global pom.xml version

def extract_cdc_version():
    with open('./pom.xml') as _f:
        return parse(_f).getroot().find(f'{XML_HEADER}properties').find(f'{XML_HEADER}revision').text


CDC_VERSION = extract_cdc_version()
print("Checking CDC version:", CDC_VERSION)

suspicious_licenses = []
excluded_packages = {'flink-cdc-source-e2e-tests', 'flink-cdc-pipeline-e2e-tests'}


def check_module_license(path, depth=0):
    jar_file = osp.join(path, 'target', f'{osp.basename(path)}-{CDC_VERSION}.jar')

    if osp.basename(path) in excluded_packages:
        # skip excluded jars
        return
    if osp.isfile(jar_file):
        check_jar_license(jar_file, depth)

    with open(osp.join(path, 'pom.xml')) as _f:
        submodules = parse(_f).getroot().find(f'{XML_HEADER}modules')
        if submodules is None:
            return
        for module in submodules:
            check_module_license(osp.join(path, module.text), depth + 1)




def check_jar_license(jar_file, depth):
    print(' ' * depth + "Checking jar", jar_file)
    with zipfile.ZipFile(jar_file) as jar:
        for name in jar.namelist():

            data = jar.read(name)
            print(name, len(data), repr(data[:10]))



if __name__ == '__main__':
    check_module_license('.')

File: tools/ci/license_check.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'rubyzip'

# Extract Flink CDC revision number from global pom.xml
begin
  REVISION_NUMBER = File.read('pom.xml').scan(%r{<revision>(.*)</revision>}).last[0]
rescue NoMethodError
  abort 'Could not extract Flink CDC revision number from pom.xml'
end

puts "Flink CDC version: '#{REVISION_NUMBER}'"

# These packages don't need to be checked
EXCLUDED_PACKAGES = %w[flink-cdc-dist flink-cdc-e2e-tests].freeze


# Traversing maven module in given path
def traverse_module(path)
  module_name = File.basename path
  return if EXCLUDED_PACKAGES.include?(module_name)

  jar_file = File.join path, 'target', "#{module_name}-#{REVISION_NUMBER}.jar"
  check_jar_license jar_file if File.exist? jar_file

  File.read(File.join(path, 'pom.xml')).scan(%r{<module>(.*)</module>}).map(&:first).each do |submodule|
    traverse_module File.join(path, submodule.to_s) unless submodule.nil?
  end
end

# Check license issues in given jar file
def check_jar_license(jar_file)
  puts "Checking jar file #{jar_file}"
end

traverse_module '.'
